### ISSUE TITLE
feat: [FOLLOW] upgrade from starknet RPC spec 0.3 to RPC spec 0.4 

### ISSUE THREAD

**Issue by Eikix** [kkrt-labs/kakarot-rpc#364]
# Feature Request

**Describe the Feature Request**

Pathfinder recently posted:
```
pathfinder v0.7.0 released

Support for Starknet v0.12.1 and RPC v0.4.0! 

[Release notes](https://github.com/eqlabs/pathfinder/releases/tag/v0.7.0)
```

**Describe Preferred Solution**

This means starknet-rs will soon support RPC v0.4.0 and it'll be wise to upgrade our types and logic to match the latest version of both.

**Comment by Eikix** [comment]
awaiting 0.4 to be released

**Comment by Eikix** [comment]
Closed in favour of #528
